### PR TITLE
`General`: Replace the remaining markdown editors with Monaco

### PR DESCRIPTION
**Pull request by pzdr7**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.

### Motivation and Context

We want to remove Ace from Artemis. However, the Markdown Editor (except for communication and programming exercise problem statements) still uses Ace.

### Description

Replaced every usage of the old markdown editor with Monaco:
- Updating text, file upload, and modeling exercises
- Legal documents (e.g. imprint)
- Standardized competencies and course competencies/prerequisites
- Exams and exam announcements
- Lectures and lecture text units
- Grading instructions (/ structured grading criteria)
- DnD and MC quiz questions

Added editor actions for the grading instructions (criterion, instruction, credits, description, scale, feedback, and usage count) and multiple choice / dnd answers (wrong, correct, hint, explanation). These generally follow the same logic of the original markdown editor.

**Note**: I will remove the remaining references to Ace / Commands / EditorMode ... in a follow-up PR where I remove Ace from Artemis.

#### Default markdown editor

Prerequisites:
- 1 Admin

Test the markdown editors. In particular, verify that your changes to the markdown text are saved and that you can see your changes in the preview.

**Exercises**

Update/Create...
- A text exercise
- A file upload exercise
- A modeling exercise

**Administrator**

Modify...
- The imprint (go to Server Administration > Imprint)
- Knowledge areas (click the 👁️ icon to edit) of standardized competencies (Server Administration > Standardized competencies) and standardized competencies

**Exam**

- Update/Create an exam
- Make a live announcement during an exam

**Courses and competencies**

- Update a course
- Update a competency in a course
- Update a lecture in a course
- Update a lecture in a course in guided mode
- Update text unit of a lecture

----

#### Special markdown editors

**Grading instructions**

Prerequisites:
- 1 Instructor

1. Create a new exercise (Text is the simplest). At the bottom of the page, add some grading instructions.
2. Check that the changes are reflected in the markdown (under "Edit as Text")
3. Check that changes in the markdown are reflected in the "Edit" view.
4. Save the exercise. Verify that your changes are saved.

**Quiz exercises (DnD; MC)**

Prerequisites:
- 1 Instructor
- 1 Student

1. As the instructor, create a new quiz. Set the quiz duration to 60-90s for convenience
5. Add a multiple choice question
6. Modify the question. Add answer options and add explanations / hints to the answer options. Add a hint or an explanation to an answer option
7. Try out the visual editor and see that the changes carry over to the markdown text
8. Add a drag and drop question
9. Add a background image and text items.
10. Modify the question's markdown text, e.g. by adding an explanation and changing the hint.
11. Save the quiz and click "Start quiz" 
12. Participate in the quiz as a student
13. Check that the questions match your input (explanations are only visible after the quiz & the results are visible)
14. As the instructor, go back to the quiz and click "Re-evaluate". Check that it displays properly and e.g. set an answer option invalid.

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| markdown-editor-monaco.component.ts | 94.57% | ✅ |
| markdown-editor-parsing.helper.ts | 100% | ✅ |
| monaco-exercise-reference.action.ts | 94.73% | ✅ |
| monaco-grading-credits.action.ts | 100% | ✅ |
| monaco-grading-criterion.action.ts | 100% | ✅ |
| monaco-grading-description.action.ts | 100% | ✅ |
| monaco-grading-feedback.action.ts | 100% | ✅ |
| monaco-grading-instruction.action.ts | 100% | ✅ |
| monaco-grading-scale.action.ts | 100% | ✅ |
| monaco-grading-usage-count.action.ts | 100% | ✅ |
| monaco-editor-action.model.ts | 94.38% | ✅ |
| monaco-editor-domain-action.model.ts | 100% | ✅ |
| monaco-formula.action.ts | 100% | ✅ |
| monaco-task.action.ts | 100% | ✅ |
| monaco-test-case.action.ts | 96% | ✅ |
| monaco-correct-multiple-choice-answer.action.ts | 100% | ✅ |
| monaco-quiz-explanation.action.ts | 100% | ✅ |
| monaco-quiz-hint.action.ts | 100% | ✅ |
| monaco-wrong-multiple-choice-answer.action.ts | 100% | ✅ |
| monaco-editor-option-preset.model.ts | 100% | ✅ |
| monaco-editor-option.helper.ts | 100% | ✅ |
| monaco-editor.component.ts | 96.15% | ✅ |

### Screenshots

Examples below

<details>
  <summary>Standardized competencies</summary>

image

</details>

<details>
  <summary>Exam live announcements</summary>

image
</details>

<details>
  <summary>Grading instructions</summary>

image
</details>

<details>
  <summary>Quiz MC</summary>

image
</details>

<details>
  <summary>Quiz DND</summary>

image
</details>

## Summary by CodeRabbit

- **New Features**
	- Introduced a new markdown editor component for improved editing experience across various sections of the application.
	- Added initial height configuration for the markdown editor, enhancing user interface adaptability.

- **Bug Fixes**
	- Streamlined button enabling logic for better usability.
	- Updated event handling for markdown content changes to ensure accurate state management.

- **Tests**
	- Updated unit tests to reflect changes in the markdown editor component and its associated functionalities.

### CODE REVIEW

File: src/main/webapp/app/lecture/wizard-mode/lecture-wizard-title.component.ts [hunk 1]
```diff
@@ -11,8 +10,7 @@ export class LectureUpdateWizardTitleComponent {
     @Input() currentStep: number;
     @Input() lecture: Lecture;
 
-    domainCommandsDescription = [new KatexCommand()];
-    EditorMode = EditorMode;
+    domainActionsDescription = [new MonacoFormulaAction()];
```
[the remark this answers is not in the document]
Comment (pzdr7) [reply src/main/webapp/app/lecture/wizard-mode/lecture-wizard-title.component.ts:13]: This is fine; these occurrences will be removed when I remove Ace from Artemis

File: src/main/webapp/app/lecture/wizard-mode/lecture-wizard-title.component.html [hunk 2]
```diff
@@ -6,6 +6,6 @@ <h1><span jhiTranslate="artemisApp.lecture.wizardMode.steps.titleStepTitle"></sp
     </form>
     <div class="form-group" id="field_description">
         <label class="form-control-label" jhiTranslate="artemisApp.lecture.description" for="field_description"></label>
-        <jhi-markdown-editor class="markdown-editor" [domainCommands]="domainCommandsDescription" [(markdown)]="lecture.description" [editorMode]="EditorMode.LATEX" />
+        <jhi-markdown-editor-monaco class="markdown-editor" [domainActions]="domainActionsDescription" [(markdown)]="lecture.description" />
```
[the remark this answers is not in the document]
Comment (pzdr7) [reply src/main/webapp/app/lecture/wizard-mode/lecture-wizard-title.component.html:9]: This is fine; these occurrences will be removed when I remove Ace from Artemis

File: src/test/playwright/support/pageobjects/exercises/quiz/QuizExerciseCreationPage.ts [hunk 3]
```diff
@@ -25,8 +25,8 @@ export class QuizExerciseCreationPage {
         await this.page.locator('#score').fill(points.toString());
 
         const fileContent = await Fixtures.get('exercise/quiz/multiple_choice/question.txt');
-        const textInputField = this.page.locator('.ace_text-input');
-        await textInputField.focus();
+        const textInputField = this.page.locator('.monaco-editor');
+        await textInputField.click();
```
[the remark this answers is not in the document]
Comment (pzdr7) [reply src/test/playwright/support/pageobjects/exercises/quiz/QuizExerciseCreationPage.ts:29]: This hasn't changed/out of scope

File: src/main/webapp/app/exercises/shared/structured-grading-criterion/grading-instructions-details/grading-instructions-details.component.ts [hunk 4]
```diff
@@ -87,7 +90,7 @@ export class GradingInstructionsDetailsComponent implements OnInit, AfterContent
         this.changeDetector.detectChanges();
         this.criteria!.forEach((criterion) => {
             criterion.structuredGradingInstructions.forEach((instruction) => {
-                this.markdownEditors.get(index)!.markdownTextChange(this.generateInstructionText(instruction));
+                this.markdownEditors.get(index)!.markdown = this.generateInstructionText(instruction);
```
[the remark this answers is not in the document]
Comment (pzdr7) [reply src/main/webapp/app/exercises/shared/structured-grading-criterion/grading-instructions-details/grading-instructions-details.component.ts:93]: This hasn't changed / the number of editors corresponds to the number of instructions

File: src/main/webapp/app/exercises/shared/structured-grading-criterion/grading-instructions-details/grading-instructions-details.component.ts [hunk 5]
```diff
@@ -217,37 +213,33 @@ export class GradingInstructionsDetailsComponent implements OnInit, AfterContent
...
+                    if (instructionActions.length !== 0) {
+                        this.setParentForInstructionsWithNoCriterion(instructionActions);
                     }
-                    criteriaCommands = domainCommands.slice(endOfInstructionsCommand - 1);
-                    if (criteriaCommands.length !== 0) {
+                    criterionActions = textWithDomainActions.slice(endOfInstructionsAction - 1);
+                    if (criterionActions.length !== 0) {
                         this.instructions = []; // resets the instructions array to be filled with the instructions of the criteria
-                        this.groupInstructionsToCriteria(criteriaCommands); // creates criterion object for each criterion and their corresponding instruction objects
+                        this.groupInstructionsToCriteria(criterionActions); // creates criterion object for each criterion and their corresponding instruction objects
```
Comment (az108) [inline src/main/webapp/app/exercises/shared/structured-grading-criterion/grading-instructions-details/grading-instructions-details.component.ts:242]: I would either keep this comment or the one in line 221. Maybe merge them?

**Review by JohannesWt — COMMENTED** [review]
Code looks good in general just two small (optional) comments

File: src/main/webapp/app/exercises/shared/structured-grading-criterion/grading-instructions-details/grading-instructions-details.component.html [hunk 6]
```diff
@@ -216,22 +218,25 @@
...
+                                            <jhi-markdown-editor-monaco
                                                 #markdownEditors
                                                 class="editor col-md-8"
-                                                [metaCommands]="[]"
-                                                [defaultCommands]="[]"
-                                                [headerCommands]="[]"
-                                                [colorCommands]="[]"
+                                                [metaActions]="[]"
+                                                [defaultActions]="[]"
+                                                [headerActions]="undefined"
```
[the remark this answers is not in the document]
Comment (pzdr7) [reply src/main/webapp/app/exercises/shared/structured-grading-criterion/grading-instructions-details/grading-instructions-details.component.html:229]: I write `[]` when the input is an array (overwrites the default values that contain a certain set of actions) and `undefined` when the input is an optional object (e.g. color action; header action) whose default value I want to overwrite with `undefined`.

File: src/main/webapp/app/exercises/shared/structured-grading-criterion/grading-instructions-details/grading-instructions-details.component.ts (hunk 5 above)
Comment (pzdr7) [reply src/main/webapp/app/exercises/shared/structured-grading-criterion/grading-instructions-details/grading-instructions-details.component.ts:242]: I'll get rid of it 👍

File: src/main/webapp/app/exercises/file-upload/manage/file-upload-exercise-update.component.ts [hunk 7]
```diff
@@ -286,4 +287,6 @@ export class FileUploadExerciseUpdateComponent implements AfterViewInit, OnDestr
         });
         this.isSaving = false;
     }
+
+    protected readonly MarkdownEditorHeight = MarkdownEditorHeight;
```
[the remark this answers is not in the document]
Comment (pzdr7) [reply src/main/webapp/app/exercises/file-upload/manage/file-upload-exercise-update.component.ts:291]: I think I just forgot to remove this. It's unused

File: src/main/webapp/app/admin/standardized-competencies/standardized-competency-management.component.ts [hunk 8]
```diff
@@ -63,6 +64,7 @@ export class StandardizedCompetencyManagementComponent extends StandardizedCompe
         private alertService: AlertService,
         private modalService: NgbModal,
         private translateService: TranslateService,
+        private cdRef: ChangeDetectorRef,
```
[the remark this answers is not in the document]
Comment (pzdr7) [reply src/main/webapp/app/admin/standardized-competencies/standardized-competency-management.component.ts:67]: Good idea 👍

File: src/main/webapp/app/admin/standardized-competencies/standardized-competency-management.component.ts [hunk 9]
```diff
@@ -85,6 +87,7 @@ export class StandardizedCompetencyManagementComponent extends StandardizedCompe
                 });
 
                 this.isLoading = false;
+                this.changeDetectorRef.detectChanges();
```
[the remark this answers is not in the document]
Comment (pzdr7) [reply src/main/webapp/app/admin/standardized-competencies/standardized-competency-management.component.ts:90]: We did it for clarity

File: src/test/javascript/spec/component/shared/monaco-editor/monaco-editor-grading-instructions.integration.spec.ts [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+        '\n\t[maxCountInScore] 0';
+
+    const generalInstructionText = 'These are some general instructions for the tutors.';
+
+    it('should insert grading instructions', () => {
+        comp.triggerKeySequence(generalInstructionText);
+        const actions = setupActions();
+        actions.instructionAction.executeInCurrentEditor();
+        expect(comp.getText()).toBe(generalInstructionText + expectedInstructionTextWithoutCriterion);
+    });
```
[the remark this answers is not in the document]
Comment (pzdr7) [reply src/test/javascript/spec/component/shared/monaco-editor/monaco-editor-grading-instructions.integration.spec.ts:66]: I don't see the benefit for combining two strings

File: src/main/webapp/app/shared/monaco-editor/model/actions/grading-criteria/monaco-grading-description.action.ts [hunk 11]
```diff
@@ -0,0 +1,20 @@
+import * as monaco from 'monaco-editor';
+import { MonacoEditorDomainAction } from 'app/shared/monaco-editor/model/actions/monaco-editor-domain-action.model';
+
+export class MonacoGradingDescriptionAction extends MonacoEditorDomainAction {
+    static readonly ID = 'monaco-grading-description.action';
+    static readonly IDENTIFIER = '[description]';
+    static readonly TEXT = 'Add grading instruction here (only visible for tutors)';
+
+    constructor() {
+        super(MonacoGradingDescriptionAction.ID, 'artemisApp.assessmentInstructions.instructions.editor.addDescription', undefined, undefined, true);
```
Comment (az108) [inline src/main/webapp/app/shared/monaco-editor/model/actions/grading-criteria/monaco-grading-description.action.ts:10]: why do we pass two undefined values here?

Comment (pzdr7) [reply src/main/webapp/app/shared/monaco-editor/model/actions/grading-criteria/monaco-grading-description.action.ts:10]: The action has no icon and no keybindings -> `undefined`

Comment (az108) [reply src/main/webapp/app/shared/monaco-editor/model/actions/grading-criteria/monaco-grading-description.action.ts:10]: alright 👍

**Review by az108 — APPROVED** [review]
Code LGTM, thank you for the quick changes 👍

**Review by JohannesWt — APPROVED** [review]
Code 👍

**Review by SimonEntholzer — APPROVED** [review]
Tested markdown editors in imprint, exam and quiz exercises on TS3

**Review by coolchock — APPROVED** [review]
tested in testing session on ts3, works as expected 👍

**Pull request merged**